File: December 2024 Leetcode Solution/December-13.java
class Solution {
    public long findScore(int[] nums) {
        int n = nums.length;
        long ans = 0;
        boolean[] visited = new boolean[n];
        List<int[]> indexedNums = new ArrayList<>();

        for (int i = 0; i < n; i++) {
            indexedNums.add(new int[]{nums[i], i});
        }

        indexedNums.sort((a, b) -> a[0] == b[0] ? a[1] - b[1] : a[0] - b[0]);

        for (int[] pair : indexedNums) {
            int value = pair[0], idx = pair[1];
            if (!visited[idx]) {
                ans += value;
                visited[idx] = true;
                if (idx > 0) visited[idx - 1] = true;
                if (idx < n - 1) visited[idx + 1] = true;
            }
        }

        return ans;
    }
}

File: December 2024 Leetcode Solution/December-13.c
typedef struct {
    int value;
    int index;
} Pair;

int compare(const void* a, const void* b) {
    Pair* p1 = (Pair*)a;
    Pair* p2 = (Pair*)b;
    if (p1->value == p2->value)
        return p1->index - p2->index;
    return p1->value - p2->value;
}

long long findScore(int* nums, int numsSize) {
    Pair* indexedNums = (Pair*)malloc(numsSize * sizeof(Pair));
    for (int i = 0; i < numsSize; i++) {
        indexedNums[i].value = nums[i];
        indexedNums[i].index = i;
    }

    qsort(indexedNums, numsSize, sizeof(Pair), compare);

    bool* vis = (bool*)calloc(numsSize, sizeof(bool));
    long long ans = 0;

    for (int i = 0; i < numsSize; i++) {
        int idx = indexedNums[i].index;
        if (!vis[idx]) {
            ans += indexedNums[i].value;
            vis[idx] = true;
            if (idx > 0) vis[idx - 1] = true;
            if (idx < numsSize - 1) vis[idx + 1] = true;
        }
    }

    free(indexedNums);
    free(vis);
    return ans;
}
